File: scripts/multi_agent_collab.py
#!/usr/bin/env python3
"""
Coordinate a round-robin, multi-agent workflow using the shared todo list.

Each agent reads the shared context, writes a note to the next agent,
and updates task status to keep ownership unambiguous.
"""

from __future__ import annotations

import argparse
from dataclasses import dataclass
from pathlib import Path
from typing import Dict, List, Optional

REPO_ROOT = Path(__file__).resolve().parent.parent
DEFAULT_CYCLES = 2


@dataclass
class TaskEntry:
    """Represents a single TODO item tracked by the collaboration loop."""

    section: str
    description: str
    owner: Optional[str] = None
    status: str = "pending"

    def key(self) -> str:
        """Return a unique key for dictionary lookups."""
        return f"{self.section} :: {self.description}"


@dataclass
class AgentNote:
    """Structured note exchanged between agents."""

    from_agent: str
    to_agent: str
    message: str


class CollaborationContext:
    """Holds shared state for all collaborating agents."""

    def __init__(self) -> None:
        # Capture repository overview for architectural notes.
        self.repo_dirs: List[str] = self._safe_scan_repo()
        # Load every unchecked TODO entry to drive the workflow.
        self.task_board: Dict[str, TaskEntry] = self._safe_load_tasks()
        # Track which tasks are assigned to each agent.
        self.assignments: Dict[str, List[str]] = {}
        # Persist the chronological log of cross-agent notes.
        self.execution_log: List[AgentNote] = []

    def _safe_scan_repo(self) -> List[str]:
        """Return top-level directories for situational awareness."""
        try:
            # Collect only directory names to avoid noisy file listings.
            dirs = [
                path.name for path in REPO_ROOT.iterdir() if path.is_dir()
            ]
            # Sort for deterministic output across runs.
            return sorted(dirs)
        except OSError as error:
            # Surface a readable message when the scan fails.
            return [f"Scan failed: {error}"]

    def _safe_load_tasks(self) -> Dict[str, TaskEntry]:
        """Parse todo.md and build a lookup keyed by section + description."""
        todo_path = REPO_ROOT / "todo.md"
        tasks: Dict[str, TaskEntry] = {}
        current_section = "General"
        try:
            # Read the entire file once; fall back to an empty plan if missing.
            contents = todo_path.read_text(encoding="utf-8")
        except FileNotFoundError:
            # No todo file means no actionable tasks today.
            return tasks
        except OSError as error:
            # Log the error as a pseudo-task so it surfaces in the workflow.
            failure = TaskEntry(
                section="System",
                description=f"Failed to read todo.md ({error})",
            )
            tasks[failure.key()] = failure
            return tasks
        # Iterate line-by-line so section headers contextualize their tasks.
        for raw_line in contents.splitlines():
            line = raw_line.strip()
            if line.startswith("##"):
                # Update the active section to group subsequent tasks.
                current_section = line.lstrip("#").strip()
                continue
            if line.startswith("- [ ]"):
                # Normalize the description and store the new task.
                description = line.replace("- [ ]", "", 1).strip()
                entry = TaskEntry(section=current_section, description=description)
                tasks[entry.key()] = entry
        return tasks

    def pending_tasks(self) -> List[str]:
        """Return every task that still needs attention."""
        return [
            key
            for key, entry in self.task_board.items()
            if entry.status not in {"done", "released"}
        ]

    def assign_task(self, key: str, owner: str) -> None:
        """Assign a task to a specific agent."""
        entry = self.task_board.get(key)
        if entry is None:
            # Ignore unknown keys so the workflow keeps flowing.
            return
        # Record the new owner and keep a per-agent assignment list.
        entry.owner = owner
        entry.status = "assigned"
        self.assignments.setdefault(owner, []).append(key)

    def update_status(self, key: str, status: str) -> None:
        """Update the lifecycle status for a tracked task."""
        entry = self.task_board.get(key)
        if entry is None:
            # Missing keys are silently ignored to keep the run resilient.
            return
        entry.status = status

    def tasks_for_agent(self, owner: str) -> List[str]:
        """Return every task assigned to the requested owner."""
        return self.assignments.get(owner, [])

    def section_dependencies(self, section: str) -> List[str]:
        """Return test tasks that gate deployments within the same section."""
        dependencies: List[str] = []
        for key, entry in self.task_board.items():
            if entry.section == section and "test" in entry.description.lower():
                dependencies.append(key)
        return dependencies

    def is_converged(self) -> bool:
        """Check whether the workflow reached a terminal state."""
        return not self.pending_tasks()


class Agent:
    """Base class shared by every specialized agent."""

    def __init__(self, name: str, recipient: str) -> None:
        self.name = name
        self.recipient = recipient

    def run(self, context: CollaborationContext) -> AgentNote:
        """Execute the agent-specific routine and record its note."""
        message = self.perform(context)
        note = AgentNote(self.name, self.recipient, message)
        context.execution_log.append(note)
        return note

    def perform(self, context: CollaborationContext) -> str:
        """Perform the agent's unique workflow."""
        raise NotImplementedError


class LeadArchitect(Agent):
    """Scans the repository and briefs the architect."""

    def perform(self, context: CollaborationContext) -> str:
        # Build a compact directory overview for the downstream architect.
        preview = ", ".join(context.repo_dirs[:6])
        if len(context.repo_dirs) > 6:
            preview += ", ..."
        # Count pending tasks to quantify workload.
        task_count = len(context.pending_tasks())
        return (
            f"Architect, repo scan highlights {task_count} pending items. "
            f"Key directories: {preview}. Please decompose the workstream."
        )


class Architect(Agent):
    """Transforms open tasks into agent-specific assignments."""

    FRONTEND_KEYWORDS = ("sidebar", "link", "ui")

    def perform(self, context: CollaborationContext) -> str:
        assignments: List[str] = []
        # Route each pending task exactly once to avoid duplicated effort.
        for key in context.pending_tasks():
            entry = context.task_board[key]
            description = entry.description.lower()
            if any(keyword in description for keyword in self.FRONTEND_KEYWORDS):
                owner = "Frontend Agent"
            elif "deploy" in description:
                owner = "Release Engineer"
            elif "test" in description:
                owner = "Backend Agent"
            else:
                owner = "Optimization Agent"
            context.assign_task(key, owner)
            assignments.append(f"{owner} ← {key}")
        if not assignments:
            return (
                "Backend Agent, no new items surfaced. Hold position until QA "
                "flags a regression."
            )
        summary = "; ".join(assignments)
        return (
            f"Backend Agent, assignments dispatched ({summary}). "
            "Kick off execution and keep the chain moving."
        )


class BackendAgent(Agent):
    """Executes backend-focused tasks before handing off to frontend."""

    def perform(self, context: CollaborationContext) -> str:
        tasks = context.tasks_for_agent(self.name)
        if not tasks:
            return (
                "Frontend Agent, backend layer is quiet. Proceed with any UI "
                "verifications needed."
            )
        completed: List[str] = []
        # Iterate through every assigned backend task and prepare it for QA.
        for key in tasks:
            entry = context.task_board[key]
            context.update_status(key, "ready-for-qa")
            completed.append(f"{entry.section} → {entry.description}")
        highlights = "; ".join(completed)
        return (
            "Frontend Agent, backend validations completed "
            f"({highlights}). Awaiting QA confirmation after your pass."
        )


class FrontendAgent(Agent):
    """Validates UI-oriented tasks and signals readiness downstream."""

    def perform(self, context: CollaborationContext) -> str:
        tasks = context.tasks_for_agent(self.name)
        if not tasks:
            return (
                "Optimization Agent, UI remains stable. No new widgets need "
                "attention."
            )
        verified: List[str] = []
        for key in tasks:
            entry = context.task_board[key]
            context.update_status(key, "done")
            verified.append(f"{entry.section} → {entry.description}")
        summary = "; ".join(verified)
        return (
            "Optimization Agent, UI verification finished "
            f"({summary}). Performance review next."
        )


class OptimizationAgent(Agent):
    """Ensures that no redundant work enters the pipeline."""

    def perform(self, context: CollaborationContext) -> str:
        # Check for any tasks that somehow remain unassigned.
        orphaned = [
            key
            for key in context.pending_tasks()
            if context.task_board[key].owner is None
        ]
        if orphaned:
            # Flag the anomaly so QA knows to pause.
            return (
                "QA Agent, detected unowned tasks "
                f"({', '.join(orphaned)}). Please hold until reassignment."
            )
        return (
            "QA Agent, throughput looks clean—no duplicated effort. "
            "Feel free to validate backend deliverables."
        )


class QAAgent(Agent):
    """Validates backend deliverables before release."""

    def perform(self, context: CollaborationContext) -> str:
        # Gather every task that the backend marked as ready-for-qa.
        ready = [
            key
            for key, entry in context.task_board.items()
            if entry.status == "ready-for-qa"
        ]
        if not ready:
            return (
                "Release Engineer, nothing new in QA. Standing by for the next "
                "drop."
            )
        validated: List[str] = []
        for key in ready:
            entry = context.task_board[key]
            context.update_status(key, "done")
            validated.append(f"{entry.section} → {entry.description}")
        report = "; ".join(validated)
        return (
            "Release Engineer, QA approved the backend scope "
            f"({report}). Prep production packages."
        )


class ReleaseEngineer(Agent):
    """Prepares production output once QA certifies dependencies."""

    def perform(self, context: CollaborationContext) -> str:
        tasks = context.tasks_for_agent(self.name)
        if not tasks:
            return (
                "Documentation Agent, no deployments queued. "
                "Capture current status anyway."
            )
        released: List[str] = []
        blocked: List[str] = []
        for key in tasks:
            entry = context.task_board[key]
            dependencies = context.section_dependencies(entry.section)
            # Ensure every dependency finished QA before releasing.
            gated = all(
                context.task_board[dep].status == "done" for dep in dependencies
            )
            if gated:
                context.update_status(key, "released")
                released.append(f"{entry.section} → {entry.description}")
            else:
                blocked.append(f"{entry.section} → waiting on tests")
        if blocked:
            details = "; ".join(blocked)
            return (
                "Documentation Agent, deployments partially blocked "
                f"({details}). Recording interim notes."
            )
        summary = "; ".join(released)
        return (
            "Documentation Agent, packaged production artifacts "
            f"({summary}). Please update the changelog."
        )


class DocumentationAgent(Agent):
    """Updates documentation based on the release payload."""

    def perform(self, context: CollaborationContext) -> str:
        # Capture every task that already reached a stable state.
        completed = [
            key
            for key, entry in context.task_board.items()
            if entry.status in {"done", "released"}
        ]
        if not completed:
            return (
                "Review Architect, documentation unchanged pending releases."
            )
        notes = "; ".join(completed)
        return (
            "Review Architect, release notes drafted covering "
            f"{notes}. Please run the final approval."
        )


class ReviewArchitect(Agent):
    """Confirms convergence and closes the collaboration loop."""

    def perform(self, context: CollaborationContext) -> str:
        # Identify any lingering tasks before closing the loop.
        remaining = context.pending_tasks()
        if remaining:
            outstanding = "; ".join(remaining)
            return (
                "Lead Architect, outstanding items detected "
                f"({outstanding}). Initiating another cycle."
            )
        return (
            "Lead Architect, all tasks converged. Ready to archive the run."
        )


def run_round_robin(cycles: int) -> CollaborationContext:
    """Execute the workflow for the requested number of cycles."""
    context = CollaborationContext()
    agents: List[Agent] = [
        LeadArchitect("Lead Architect", "Architect"),
        Architect("Architect", "Backend Agent"),
        BackendAgent("Backend Agent", "Frontend Agent"),
        FrontendAgent("Frontend Agent", "Optimization Agent"),
        OptimizationAgent("Optimization Agent", "QA Agent"),
        QAAgent("QA Agent", "Release Engineer"),
        ReleaseEngineer("Release Engineer", "Documentation Agent"),
        DocumentationAgent("Documentation Agent", "Review Architect"),
        ReviewArchitect("Review Architect", "Lead Architect"),
    ]
    # Run the requested number of cycles or stop early when converged.
    for _ in range(max(1, cycles)):
        for agent in agents:
            agent.run(context)
        if context.is_converged():
            break
    return context


def print_log(context: CollaborationContext) -> None:
    """Pretty-print the agent log and the final task statuses."""
    print("Multi-Agent Collaboration Log")
    print("-" * 40)
    for note in context.execution_log:
        print(f"{note.from_agent} → {note.to_agent}: {note.message}")
    print("\nTask Statuses")
    print("-" * 40)
    for key, entry in context.task_board.items():
        print(f"[{entry.status.upper():>8}] {key}")


def parse_args() -> argparse.Namespace:
    """Parse CLI arguments."""
    parser = argparse.ArgumentParser(
        description="Activate the autonomous multi-agent collaboration loop."
    )
    parser.add_argument(
        "--cycles",
        type=int,
        default=DEFAULT_CYCLES,
        help="Number of full round-robin cycles to execute.",
    )
    return parser.parse_args()


def main() -> None:
    """Entrypoint used when running as a script."""
    args = parse_args()
    context = run_round_robin(cycles=args.cycles)
    print_log(context)


if __name__ == "__main__":
    main()

File: scripts/multiAgentCollab.ts
import fs from "fs";
import path from "path";

/**
 * Represents the canonical name of every collaborating agent in the workflow.
 */
type AgentName =
  | "Lead Architect"
  | "Architect"
  | "Backend Engineer"
  | "Frontend Engineer"
  | "Optimization Engineer"
  | "QA Agent"
  | "Release Engineer"
  | "Documentation Agent";

/**
 * Captures a single TODO item alongside its inferred assignee and section.
 */
interface TodoTask {
  section: string;
  text: string;
  assignee: AgentName | "Architect";
}

/**
 * Describes the structure of the repository-level statistics that the lead architect reports.
 */
interface RepoStats {
  topLevelEntries: number;
  docsFileCount: number;
  srcFileCount: number;
}

/**
 * Represents a note sent from one agent to another.
 */
interface AgentActionLog {
  from: string;
  to: string;
  note: string;
}

const repoRoot = path.resolve(__dirname, "..");
const docsDir = path.join(repoRoot, "docs");
const todoPath = path.join(repoRoot, "todo.md");
const logPath = path.join(docsDir, "multi-agent-collab-log.md");

/**
 * Safely reads a file and falls back to an empty string on failure so the automation can continue.
 *
 * @param filePath - Absolute path to the file that should be read.
 * @returns The file contents or an empty string when the file is missing.
 */
function safeReadFile(filePath: string): string {
  try {
    // Attempt to read the requested file synchronously to keep the script deterministic.
    return fs.readFileSync(filePath, "utf-8");
  } catch {
    // Return an empty string if the file does not exist so downstream parsing can handle the absence.
    return "";
  }
}

/**
 * Recursively counts every file beneath a directory.
 *
 * @param targetDir - The directory whose file count should be calculated.
 * @returns The number of files found underneath the directory.
 */
function countFiles(targetDir: string): number {
  // Guard against directories that do not exist to prevent runtime errors.
  if (!fs.existsSync(targetDir)) {
    return 0;
  }

  // Walk the directory tree depth-first to tally every file.
  return fs.readdirSync(targetDir).reduce((count, entry) => {
    const entryPath = path.join(targetDir, entry);
    const stats = fs.statSync(entryPath);

    if (stats.isDirectory()) {
      // Recurse into child directories to include their files in the total.
      return count + countFiles(entryPath);
    }

    // Increment the count for every file discovered at the current level.
    return count + 1;
  }, 0);
}

/**
 * Generates high-level repository statistics for the Lead Architect.
 *
 * @returns Counts describing top-level entries, docs files, and src files.
 */
function collectRepoStats(): RepoStats {
  // List the top-level directory to understand project breadth.
  const topLevelEntries = fs.readdirSync(repoRoot).length;
  // Count Markdown files in the docs directory to measure documentation scale.
  const docsFileCount = fs.existsSync(docsDir) ? fs.readdirSync(docsDir).length : 0;
  // Recursively count source files to quantify implementation surface area.
  const srcFileCount = countFiles(path.join(repoRoot, "src"));

  return { topLevelEntries, docsFileCount, srcFileCount };
}

/**
 * Assigns TODO items to the appropriate agents based on simple heuristics.
 *
 * @param text - The textual content of a TODO entry.
 * @returns The agent that should tackle the task.
 */
function determineAssignee(text: string): AgentName | "Architect" {
  // Normalize the text for easier keyword comparisons.
  const lower = text.toLowerCase();

  if (lower.includes("test") || lower.includes("verify")) {
    return "QA Agent";
  }

  if (lower.includes("deploy") || lower.includes("release")) {
    return "Release Engineer";
  }

  if (
    lower.includes("ui") ||
    lower.includes("link") ||
    lower.includes("page") ||
    lower.includes("frontend")
  ) {
    return "Frontend Engineer";
  }

  if (
    lower.includes("api") ||
    lower.includes("service") ||
    lower.includes("github") ||
    lower.includes("backend")
  ) {
    return "Backend Engineer";
  }

  if (lower.includes("optimiz") || lower.includes("performance")) {
    return "Optimization Engineer";
  }

  // Fall back to the architect when a task does not clearly belong to another specialist.
  return "Architect";
}

/**
 * Parses the shared TODO markdown file into structured tasks.
 *
 * @param markdown - The raw contents of todo.md.
 * @returns A list of structured TODO tasks complete with inferred assignees.
 */
function parseTodoTasks(markdown: string): TodoTask[] {
  const tasks: TodoTask[] = [];
  const lines = markdown.split("\n");
  let currentSection = "General";

  // Iterate line-by-line to track the latest section heading.
  for (const rawLine of lines) {
    const line = rawLine.trim();

    if (line.startsWith("## ")) {
      // Update the section when a heading is encountered.
      currentSection = line.replace(/^##\s*/, "").trim();
      continue;
    }

    if (line.startsWith("- [ ]")) {
      // Remove the checkbox syntax to isolate the actionable text.
      const text = line.replace("- [ ]", "").trim();
      tasks.push({
        section: currentSection,
        text,
        assignee: determineAssignee(text),
      });
    }
  }

  return tasks;
}

/**
 * Groups tasks by their inferred assignee for quick lookup during the round-robin loop.
 *
 * @param tasks - Structured TODO entries awaiting assignment.
 * @returns A mapping of agent names to the tasks they should execute.
 */
function groupTasksByAssignee(tasks: TodoTask[]): Record<string, TodoTask[]> {
  const bucket: Record<string, TodoTask[]> = {
    "Lead Architect": [],
    Architect: [],
    "Backend Engineer": [],
    "Frontend Engineer": [],
    "Optimization Engineer": [],
    "QA Agent": [],
    "Release Engineer": [],
    "Documentation Agent": [],
  };

  // Drop each task into the matching bucket.
  tasks.forEach((task) => {
    const key = bucket[task.assignee] ? task.assignee : "Architect";
    bucket[key].push(task);
  });

  return bucket;
}

/**
 * Formats a list of tasks into a concise human-readable string.
 *
 * @param tasks - Tasks that belong to a given agent.
 * @returns A bullet string summarizing the assignments.
 */
function formatTaskList(tasks: TodoTask[]): string {
  if (!tasks.length) {
    return "No open items in my lane. Standing by to support the next specialist.";
  }

  // Join the tasks with semicolons to keep the note compact.
  return tasks
    .map((task) => `${task.section}: ${task.text}`)
    .join("; ");
}

/**
 * Writes a Markdown log capturing the collaboration cycle.
 *
 * @param stats - Repository statistics shared by the lead architect.
 * @param tasks - All parsed tasks for reference.
 * @param conversation - Notes exchanged between agents.
 */
function writeLog(stats: RepoStats, tasks: TodoTask[], conversation: AgentActionLog[]): void {
  // Ensure the docs directory exists to avoid write errors.
  if (!fs.existsSync(docsDir)) {
    fs.mkdirSync(docsDir, { recursive: true });
  }

  const timestamp = new Date().toISOString();
  const header = `## ${timestamp}\n`;
  const statsBlock = [
    `- Top-level entries: ${stats.topLevelEntries}`,
    `- docs/ files: ${stats.docsFileCount}`,
    `- src/ files: ${stats.srcFileCount}`,
    `- Open TODO items: ${tasks.length}`,
  ].join("\n");

  const conversationBlock = conversation
    .map((entry, index) => `${index + 1}. [${entry.from} ➜ ${entry.to}] ${entry.note}`)
    .join("\n");

  const body = `${header}${statsBlock}\n\n### Conversation\n${conversationBlock}\n\n`;

  // Append the new log entry so prior collaboration history is preserved.
  fs.appendFileSync(logPath, body, "utf-8");
}

/**
 * Executes the scripted collaboration loop and returns the generated conversation.
 *
 * @param groupedTasks - Tasks grouped by assignee.
 * @param stats - Repository statistics for context.
 * @returns The ordered list of agent notes.
 */
function runCollaboration(
  groupedTasks: Record<string, TodoTask[]>,
  stats: RepoStats,
): AgentActionLog[] {
  const sequence: AgentActionLog[] = [];

  // Lead Architect hands context to the Architect.
  sequence.push({
    from: "Lead Architect",
    to: "Architect",
    note: `Scanned ${stats.topLevelEntries} top-level entries, ${stats.docsFileCount} docs files, and ${stats.srcFileCount} source files. No structural drift detected.`,
  });

  // Architect sets the agenda based on parsed TODOs.
  const totalTasks = Object.values(groupedTasks).reduce((sum, list) => sum + list.length, 0);
  sequence.push({
    from: "Architect",
    to: "Backend Engineer",
    note: `Prioritized ${totalTasks} open items. Handing off service-facing fixes first to unblock downstream roles.`,
  });

  // Backend Engineer summarizes their lane.
  sequence.push({
    from: "Backend Engineer",
    to: "Frontend Engineer",
    note: formatTaskList(groupedTasks["Backend Engineer"]),
  });

  // Frontend Engineer shares UI work status.
  sequence.push({
    from: "Frontend Engineer",
    to: "Optimization Engineer",
    note: formatTaskList(groupedTasks["Frontend Engineer"]),
  });

  // Optimization Engineer addresses performance considerations.
  sequence.push({
    from: "Optimization Engineer",
    to: "QA Agent",
    note: formatTaskList(groupedTasks["Optimization Engineer"]),
  });

  // QA Agent highlights pending validation steps.
  sequence.push({
    from: "QA Agent",
    to: "Release Engineer",
    note: formatTaskList(groupedTasks["QA Agent"]),
  });

  // Release Engineer prepares deployment guidance.
  sequence.push({
    from: "Release Engineer",
    to: "Documentation Agent",
    note: formatTaskList(groupedTasks["Release Engineer"]),
  });

  // Documentation Agent records the cycle summary.
  sequence.push({
    from: "Documentation Agent",
    to: "Architect",
    note: "Log drafted in docs/multi-agent-collab-log.md with the latest conversation and metrics.",
  });

  // Architect performs the final review for convergence.
  sequence.push({
    from: "Architect",
    to: "All",
    note: "Cycle complete. Pending items remain in QA and Release lanes; no duplicate work detected.",
  });

  return sequence;
}

/**
 * Entry point that wires together every helper and prints the summary to stdout.
 */
function main(): void {
  // Gather repository context for the opening note.
  const stats = collectRepoStats();
  // Parse TODO entries to understand outstanding work.
  const todoMarkdown = safeReadFile(todoPath);
  const tasks = parseTodoTasks(todoMarkdown);
  // Group tasks per agent to respect ownership boundaries.
  const groupedTasks = groupTasksByAssignee(tasks);
  // Run the collaboration loop and capture the conversation.
  const conversation = runCollaboration(groupedTasks, stats);
  // Persist the latest cycle into documentation.
  writeLog(stats, tasks, conversation);

  // Output a concise summary for interactive use.
  console.log("Multi-agent collaboration cycle complete.");
  console.table(
    Object.entries(groupedTasks).map(([agent, items]) => ({
      agent,
      openItems: items.length,
    })),
  );
}

// Execute the script if it is invoked directly.
main();
